src/uct/cuda/cuda_copy/cuda_copy_md.c

```diff
@@ -286,74 +482,109 @@ static size_t uct_cuda_copy_md_get_total_device_mem(CUdevice cuda_device)
-         * change in the future. Ideally checking for
-         * `CU_POINTER_ATTRIBUTE_IS_LEGACY_CUDA_IPC_CAPABLE` would be better
-         * here, but due to a bug in the driver `cudaMalloc` also returns false
-         * in that case. Therefore, checking whether the allocation was not
-         * allocated in a context should also allows us to identify
-         * virtual/stream-ordered CUDA allocations. */
-        mem_info->type = UCS_MEMORY_TYPE_CUDA_MANAGED;
-
-        cu_err =
-            cuMemRangeGetAttribute((void*)&pref_loc, sizeof(pref_loc),
-                                   CU_MEM_RANGE_ATTRIBUTE_PREFERRED_LOCATION,
-                                   (CUdeviceptr)address, length);
-        if ((cu_err != CUDA_SUCCESS) || (pref_loc == CU_DEVICE_INVALID)) {
-            pref_loc = (md->config.pref_loc == UCT_CUDA_PREF_LOC_CPU) ?
-                CU_DEVICE_CPU : cuda_device;
+        if (cuda_mem_type != CU_MEMORYTYPE_DEVICE) {
+            /* pointer not recognized */
+            return UCS_ERR_INVALID_ADDR;
         }
 
-        if (pref_loc == CU_DEVICE_CPU) {
-            mem_info->sys_dev = UCS_SYS_DEVICE_ID_UNKNOWN;
-        } else {
-            uct_cuda_base_get_sys_dev(pref_loc, &mem_info->sys_dev);
-            if (mem_info->sys_dev == UCS_SYS_DEVICE_ID_UNKNOWN) {
-                ucs_diag("cuda device %d (for address %p...%p) unrecognized",
-                         pref_loc, address,
-                         UCS_PTR_BYTE_OFFSET(address, length));
+        mem_info->type = UCS_MEMORY_TYPE_CUDA;
+
+        if (is_managed) {
+            /* is_managed: cuMemGetAddress range does not support managed memory
+             * so use provided address and length as base address and alloc
+             * length respectively */
+            mem_info->type = UCS_MEMORY_TYPE_CUDA_MANAGED;
+
+            cu_err = cuMemRangeGetAttribute(
+                    (void*)&pref_loc, sizeof(pref_loc),
+                    CU_MEM_RANGE_ATTRIBUTE_PREFERRED_LOCATION,
+                    (CUdeviceptr)address, length);
+            if ((cu_err != CUDA_SUCCESS) || (pref_loc == CU_DEVICE_INVALID)) {
+                pref_loc = (md->config.pref_loc == UCT_CUDA_PREF_LOC_CPU) ?
+                                   CU_DEVICE_CPU :
+                                   cuda_device;
+            }
+
+            if (pref_loc == CU_DEVICE_CPU) {
+                mem_info->sys_dev = UCS_SYS_DEVICE_ID_UNKNOWN;
+            } else {
+                uct_cuda_base_get_sys_dev(pref_loc, &mem_info->sys_dev);
+                if (mem_info->sys_dev == UCS_SYS_DEVICE_ID_UNKNOWN) {
+                    ucs_diag("cu_device %d (for address %p...%p) unrecognized",
+                             pref_loc, address,
+                             UCS_PTR_BYTE_OFFSET(address, length));
+                }
             }
+
+            goto out_default_range;
         }
+    }
 
+    if (cuda_device == CU_DEVICE_CPU) {
```

Comment:
can we move lines 591..596 (inclusive) to be the 'else' block of this if condition?

Reply:
ok